tools/validate_spec.py

```diff
@@ -3,261 +3,286 @@
+            f'message "{str(identifier.name)}" could not be '
+            f'found in package "{str(identifier.parent)}"'
+        ) from e
+    except FileNotFoundError as e:
+        raise ValidationError(f"specification {str(e)}") from e
+
+    classified_incorrectly = 0
+    with JsonOutputWriter(full_output_path) as json_output_writer:
+        validator = Validator(pdu_message)
+        for path, path_generator in path_generators.items():
+            for message_file_path in path_generator:
+                msg = get_message_to_validate(message_file_path, path == path_valid)
+                validation_result = validator.validate_message(msg)
                 json_output_writer.write(validation_result.get_json_output())
-            print(validation_result.get_abbreviated_output())
-            if abort_on_error and validation_result.get_classification() in [
-                Classification.FP,
-                Classification.FN,
-            ]:
-                if json_output_writer is not None:
-                    json_output_writer.close()
-                sys.exit(1)
+                print(validation_result.get_abbreviated_output())
+                if validation_result.classification in [
+                    Classification.FP,
+                    Classification.FN,
+                ]:
+                    if abort_on_error:
+                        sys.exit(1)
+                    else:
+                        classified_incorrectly += 1
 
-    if json_output_writer is not None:
-        json_output_writer.close()
+    print(f"Summary: {classified_incorrectly} messages were classified incorrectly.")
+    if classified_incorrectly != 0:
+        sys.exit(1)
 
 
 class JsonOutputWriter:
-    def __init__(self, file: Path) -> None:
-        self.file = open(file, "w+")
-        self.file.write("[\n")
-        self.count = 0
+    file: Optional[IO]
+
+    def __init__(self, file: Optional[Path]) -> None:
+        if file is not None:
+            try:
+                self.file = open(file, "w")
+            except OSError as e:
+                raise ValidationError(f"cannot open output file {file}: {e}") from e
+            self.file.write("[\n")
+            self.count = 0
+        else:
+            self.file = file
 
-    def write(self, content: Dict[str, object]) -> None:
-        if self.count != 0:
-            self.file.write(",\n")
-        json.dump(
-            content,
-            self.file,
-            indent="\t",
-        )
-        self.count += 1
+    def __enter__(self) -> "JsonOutputWriter":
+        return self
 
-    def close(self) -> None:
-        self.file.write("\n]")
-        self.file.close()
+    def __exit__(
+        self,
+        exception_type: Optional[Type[BaseException]],
+        exception_value: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
 
+        if exception_value is not None:
+            raise ValidationError(str(exception_value))
```

Comment:
Why are exception without value ignored? Why are all exceptions converted to `ValidationError`? This prevents the distinction of validation errors from internal Python exceptions.

Reply:
Removed this because exceptions that are passed into `exit` should not be reraised explicitly.
dde1b330